File: ios/XiriGoEcommerce/Core/DesignSystem/Component/XGRatingBar.swift
import SwiftUI

// MARK: - XGRatingBar

/// Star rating display with optional numeric value and review count.
/// Token source: `components.json > XGStarRating`.
///
/// - Star size: 12pt
/// - Star gap: 2pt
/// - Count: 5 stars
/// - Review count spacing: 4pt from stars
/// - Review count font: 12pt, textSecondary color
struct XGRatingBar: View {
    // MARK: - Lifecycle

    init(
        rating: Double,
        maxRating: Int = Constants.starCount,
        starSize: CGFloat = Constants.starSize,
        showValue: Bool = false,
        reviewCount: Int? = nil,
    ) {
        self.rating = rating
        self.maxRating = maxRating
        self.starSize = starSize
        self.showValue = showValue
        self.reviewCount = reviewCount
    }

    // MARK: - Internal

    var body: some View {
        HStack(spacing: Constants.reviewCountSpacing) {
            starsView

            if showValue {
                Text(String(format: "%.1f", rating))
                    .font(.system(size: Constants.reviewCountFontSize))
                    .foregroundStyle(XGColors.onSurfaceVariant)
                    .lineLimit(1)
                    .fixedSize()
            }

            if let reviewCount {
                Text("(\(reviewCount))")
                    .font(.system(size: Constants.reviewCountFontSize))
                    .foregroundStyle(XGColors.onSurfaceVariant)
                    .lineLimit(1)
                    .fixedSize()
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityDescription)
    }

    // MARK: - Private

    private enum Constants {
        static let starSize: CGFloat = 12
        static let starGap: CGFloat = 2
        static let starCount = 5
        static let reviewCountFontSize: CGFloat = 12
        static let reviewCountSpacing: CGFloat = 4
    }

    private let rating: Double
    private let maxRating: Int
    private let starSize: CGFloat
    private let showValue: Bool
    private let reviewCount: Int?

    private var accessibilityDescription: String {
        var desc = String(localized: "common_rating_description \(rating) \(maxRating)")
        if let reviewCount {
            desc += " " + String(localized: "common_reviews_count \(reviewCount)")
        }
        return desc
    }

    private var starsView: some View {
        HStack(spacing: Constants.starGap) {
            ForEach(1 ... maxRating, id: \.self) { position in
                starImage(for: position)
                    .font(.system(size: starSize))
            }
        }
    }

    @ViewBuilder
    private func starImage(for position: Int) -> some View {
        let positionDouble = Double(position)
        if rating >= positionDouble {
            Image(systemName: "star.fill")
                .foregroundStyle(XGColors.ratingStarFilled)
                .accessibilityHidden(true)
        } else if rating >= positionDouble - 0.5 {
            Image(systemName: "star.leadinghalf.filled")
                .foregroundStyle(XGColors.ratingStarFilled)
                .accessibilityHidden(true)
        } else {
            Image(systemName: "star")
                .foregroundStyle(XGColors.ratingStarEmpty)
                .accessibilityHidden(true)
        }
    }
}

// MARK: - Previews

#Preview("XGRatingBar Default") {
    VStack(spacing: XGSpacing.sm) {
        XGRatingBar(rating: 4.5, showValue: true, reviewCount: 123)
        XGRatingBar(rating: 3.0, showValue: true)
        XGRatingBar(rating: 1.5)
        XGRatingBar(rating: 5.0, starSize: 24, showValue: true)
    }
    .padding()
}

File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/XGRatingBar.kt
package com.xirigo.ecommerce.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.StarHalf
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.outlined.StarOutline
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.xirigo.ecommerce.R
import com.xirigo.ecommerce.core.designsystem.theme.PoppinsFontFamily
import com.xirigo.ecommerce.core.designsystem.theme.XGColors
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

// components.json: XGStarRating
private val StarSize = 12.dp
private val StarGap = 2.dp
private const val STAR_COUNT = 5
private val ReviewCountFontSize = 12.sp
private val ReviewCountLineHeight = 16.sp
private val ReviewCountSpacing = 4.dp

/** Star rating bar with optional numeric value and review count display. */
@Composable
fun XGRatingBar(
    rating: Float,
    modifier: Modifier = Modifier,
    maxRating: Int = STAR_COUNT,
    starSize: Dp = StarSize,
    showValue: Boolean = false,
    reviewCount: Int? = null,
) {
    val ratingDescription = stringResource(R.string.common_rating_description, rating, maxRating)
    val reviewsDescription = reviewCount?.let {
        stringResource(R.string.common_reviews_count, it)
    }.orEmpty()

    Row(
        modifier = modifier.semantics {
            contentDescription = "$ratingDescription $reviewsDescription".trim()
        },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(StarGap),
    ) {
        RatingStars(rating = rating, maxRating = maxRating, starSize = starSize)
        RatingValueText(rating = rating, showValue = showValue)
        ReviewCountText(reviewCount = reviewCount)
    }
}

@Composable
private fun RatingStars(
    rating: Float,
    maxRating: Int,
    starSize: Dp,
) {
    for (i in 1..maxRating) {
        val icon = when {
            rating >= i.toFloat() -> Icons.Filled.Star
            rating >= i.toFloat() - 0.5f -> Icons.AutoMirrored.Filled.StarHalf
            else -> Icons.Outlined.StarOutline
        }

        val tint = if (rating >= i.toFloat() - 0.5f) {
            XGColors.RatingStarFilled
        } else {
            XGColors.RatingStarEmpty
        }

        Icon(
            imageVector = icon,
            contentDescription = null,
            modifier = Modifier.size(starSize),
            tint = tint,
        )
    }
}

@Composable
private fun RatingValueText(rating: Float, showValue: Boolean) {
    if (showValue) {
        Spacer(modifier = Modifier.width(ReviewCountSpacing))
        Text(
            text = String.format(java.util.Locale.ROOT, "%.1f", rating),
            fontFamily = PoppinsFontFamily,
            fontSize = ReviewCountFontSize,
            fontWeight = FontWeight.Normal,
            color = XGColors.OnSurfaceVariant,
            lineHeight = ReviewCountLineHeight,
        )
    }
}

@Composable
private fun ReviewCountText(reviewCount: Int?) {
    if (reviewCount != null) {
        Spacer(modifier = Modifier.width(ReviewCountSpacing))
        Text(
            text = stringResource(R.string.common_review_count_format, reviewCount),
            fontFamily = PoppinsFontFamily,
            fontSize = ReviewCountFontSize,
            fontWeight = FontWeight.Normal,
            color = XGColors.OnSurfaceVariant,
            lineHeight = ReviewCountLineHeight,
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun XGRatingBarPreview() {
    XGTheme {
        XGRatingBar(rating = 4.5f, showValue = true, reviewCount = 123)
    }
}

@Preview(showBackground = true)
@Composable
private fun XGRatingBarLowPreview() {
    XGTheme {
        XGRatingBar(rating = 2.0f)
    }
}

@Preview(showBackground = true)
@Composable
private fun XGRatingBarFullPreview() {
    XGTheme {
        XGRatingBar(rating = 5.0f, showValue = true)
    }
}
